Update layer list context keys

alisterburt opened the pull request:
# Description
This PR updates the context keys available on the `LayerList`, increasing the scope of what can be inferred from our `LayerList` context

- missing `only_<layer_class>_selected` context keys added
- `num_<layer_class>_selected` context keys added
- updated some keys to unify naming style

Internal usages of layer list context keys have been updated to reflect the new names. @tlambert03 assures me that these shouldn't be being used by others yet and changing the names to unify is okay at this stage.

@napari/core-devs could you take a look at these `ContextKey` names? Once these get out in the open any changes will be hard to revert 👀 

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# References
@tlambert03 + https://napari.org/guides/contexts_expressions.html

# How has this been tested?
<!-- Please describe the tests that you ran to verify your changes. -->
- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

alisterburt committed:
add context functions for number of each layer type

alisterburt committed:
add context keys

alisterburt committed:
update layer actions and contex key names

alisterburt committed:
fix test

andy-sweet reviewed (approved):
I like the PR as is, but like the summarized table from @sofroniewn even more - i.e. the smaller set of context keys with the `only*` keys.

alisterburt committed:
remove 'only_*' layer actions, replace with compound expressions

alisterburt committed:
remove 'only_*' contexts

alisterburt committed:
add num selected layers linked

alisterburt committed:
all_layers_same_shape -> all_selected_layers_same_shape

alisterburt committed:
add func for all same type

alisterburt committed:
fix type stuff

alisterburt committed:
add test for all selected layers same type key

alisterburt committed:
add guard to same type context for length 0 selections

alisterburt committed:
hmm not working '../../../layers/_layer_actions.py:274: in <module>
    LLCK.num_selected_labels_layers >= 1 & LLCK.all_selected_layers_same_type
E   TypeError: unsupported operand type(s) for &: 'int' and 'ContextKey''

alisterburt committed:
praise be to brackets

tlambert03 commented on napari/utils/context/_layerlist_context.py:
```diff
@@ -73,31 +117,42 @@ def _active_dtype(s: LayerSel) -> DTypeLike:
+        next_layer = s[idx + 1] if idx < (len(s) - 1) else None
+        if (type(layer) != type(next_layer)) and (next_layer is not None):
+            return False
+    return True
```
what about?

```suggestion
    return len({x._type_string for x in s}) == 1
```

alisterburt replied:
🤣 I love it - thanks!!

alisterburt committed:
Update napari/utils/context/_layerlist_context.py

Co-authored-by: Talley Lambert <talley.lambert@gmail.com>

Czaki reviewed (dismissed):
I do no see update of [guides/context_expressions.md](https://github.com/napari/napari/blob/e39b79bea1671b9a265e9c55316415da0ede1e19/docs/guides/contexts_expressions.md#napari-context-keys) in this PR, but this file contains table of example context keys that will be outdated after this PR merge. Could you update it? 

Is there place where all context keys are documented? or only code and PR comments (I'm not familiar with this part of napari code)?
